tools/acorn-optimizer.js

```diff
@@ -628,15 +637,68 @@ function emitDCEGraph(ast) {
               // name may be slightly different (extra "_" in wasm backend)
               saveAsmExport(name, asmName);
               emptyOut(node); // ignore this in the second pass; this does not root
+              return;
             }
           }
         }
       }
+      // A variable declaration that has no initial values can be ignored in
+      // the second pass, these are just declarations, not roots - an actual
+      // use must be found in order to root.
+      if (!node.declarations.reduce(function(hasInit, decl) {
+            return hasInit || !!decl.init
+          }, false)) {
+        emptyOut(node);
+      }
     } else if (node.type === 'FunctionDeclaration') {
       defuns.push(node);
       var name = node.id.name;
       nameToGraphName[name] = getGraphName(name, 'defun');
       emptyOut(node); // ignore this in the second pass; we scan defuns separately
+    } else if (node.type === 'FunctionExpression') {
+      // Check if this is the minimal runtime exports function, which looks like
+      //   function(output) { var asm = output.instance.exports;
```

Comment:
Hmm, does this require that the exports look like
```
function(output) { var asm = output.instance.exports;
```
?
In some cases the exports look instead like
```
var asm;
function(output) { asm = output.instance.exports;
```
(in particular `#if !(LibraryManager.has('library_exports.js') && (WASM || WASM_BACKEND)) && !EMBIND` is false)

Reply:
Oh, then maybe there's another pattern then that we should match. But sounds like it's only for fastcomp, though? I'd prefer to not do it in that case unless this is especially urgent for some reason.

Reply:
The global `var asm;` occurs 1) if using `library_exports.js` (compiling with `-lexports.js`), or 2) if using embind, or 3) if using asm.js in fastcomp. The cases 1) and 2) also happen in wasm backend.